File: src/main/scala/org/joisen/chapter11/SimpleTableExample.scala
package org.joisen.chapter11

import org.apache.flink.streaming.api.scala._
import org.apache.flink.table.api.Expressions.$
import org.apache.flink.table.api.Table
import org.apache.flink.table.api.bridge.scala.StreamTableEnvironment
import org.joisen.chapter05.Event

/**
 * @Author Joisen
 * @Date 2022/11/26 10:42
 * @Version 1.0
 */
object SimpleTableExample {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    // 读取数据，创建DataStream
    val eventStream: DataStream[Event] = env.fromElements(
      Event("Alice", "./home", 1000L),
      Event("Bob", "./cart", 1000L),
      Event("Alice", "./prod?id=1", 5 * 1000L),
      Event("Cary", "./home", 60 * 1000L),
      Event("Bob", "./prod?id=3", 90 * 1000L),
      Event("Alice", "./prod?id=7", 105 * 1000L)
    )

    // 创建表环境
    val tableEnv: StreamTableEnvironment = StreamTableEnvironment.create(env)

    // 将DataStream转换成表
    val eventTable: Table = tableEnv.fromDataStream(eventStream)

    // 调用table Api进行转换计算
    val resTable: Table = eventTable.select($("url"), $("user"))
      .where($("user").isEqual("Alice"))

    // 直接写SQL
    val resSqlTable: Table = tableEnv.sqlQuery("select url, user from " + eventTable + " where user = 'Bob'")


    // 转换成流 打印输出
    tableEnv.toDataStream(resTable).print("Table")
    tableEnv.toDataStream(resSqlTable).print("SQL")

    env.execute()
  }
}

File: src/main/java/org/joisen/java/chapter11/SimpleTableExample.java
package org.joisen.java.chapter11;

import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
import org.joisen.java.chapter05.ClickSource;
import org.joisen.java.chapter05.Event;

import java.time.Duration;

import static org.apache.flink.table.api.Expressions.$;

/**
 * @Author Joisen
 * @Date 2022/12/12 15:01
 * @Version 1.0
 */
public class SimpleTableExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        SingleOutputStreamOperator<Event> eventStream = env.addSource(new ClickSource())
                // 乱序流的watermark生成
                .assignTimestampsAndWatermarks(WatermarkStrategy.<Event>forBoundedOutOfOrderness(Duration.ZERO)
                        .withTimestampAssigner(new SerializableTimestampAssigner<Event>() {
                            @Override
                            public long extractTimestamp(Event event, long l) {
                                return event.timestamp;
                            }
                        }));
        // 创建表执行环境
        StreamTableEnvironment tableEnv = StreamTableEnvironment.create(env);
        // 将DataStream转换成Table
        Table eventTable = tableEnv.fromDataStream(eventStream);
        // 直接写SQL进行转换
        Table resTable1 = tableEnv.sqlQuery("select user, url from " + eventTable);

        // 基于Table直接转换
        Table resTable2 = eventTable.select($("user"), $("url"))
                .where($("user").isEqual("Alice"));

        tableEnv.toDataStream(resTable1).print("result1");
        tableEnv.toDataStream(resTable2).print("result2");
        env.execute();
    }
}
